README
NAME
    Database::Migrator - A system for implementing database migrations

VERSION
    version 0.11

DESCRIPTION
    This distribution consists of a single role, Database::Migrator::Core.
    This role can be consumed by classes which implement the required
    methods for the role. These classes will then implement a complete
    database schema creation and migration system.

MIGRATION ARCHITECTURE
    The migration system starts with a file containing the DDL (database
    description language) for the full database schema. If the database
    doesn't yet exist, the database will be created and this DDL will be run
    against it.

    This DDL file *should not* contain any sort of "CREATE DATABASE"
    statement. This will be done separately before the DDL is run.

    This DDL file *may* contain DDL to create users and grant them access to
    the database.

    Once the database exists, the migrations are run against the database.

    Each migration goes into its own directory. The directory name is the
    name of the migration. Migrations are applied in sorted order. If the
    migrations start with numbers, they are sorted by these numbers,
    otherwise they are sorted alphabetically.

    The migration directory can either contain files with SQL or Perl. If a
    file ends in ".sql", the migration runner code will feed it to the
    appropriate command line utility for your database.

    Otherwise the file is assumed to contain Perl code. This code is
    expected to return a single anonymous subroutine when "eval"ed. This
    subroutine will then be called with the "Database::Migrator" object as
    its only argument.

    Each file in a single migration's directory is run in sorted order. You
    can use numeric prefixes on these files if necessary.

    Once a migration has been applied, that fact is stored in the database,
    and the migration will not be applied again. This is done by recording
    the migration's name in a table. The name of the table is determined by
    your code. I recommend something like "AppliedMigration" or
    "applied_migrations", depending on your table naming scheme.

  Migration Example
    Let's assume a set of files like this:

      migrations/
       |
       |-- 01-add-foo-data/
       |     \
       |      \-- 01-create-foo-table.sql
       |      |
       |      |-- 02-insert-foo-data.pl
       |
       |-- 02-add-bar-table/
            \
             \-- add-bar-table.sql

    The 01-add-foo-data/01-create-foo-table.sql file might look like this:

      CREATE TABLE Foo (
        foo_id       INT   PRIMARY KEY AUTO_INCREMENT,
        size         INT   NOT NULL,
        description  TEXT  NOT NULL
      );

    The 01-add-foo-data/02-insert-foo-data.pl file might contain this:

      sub {
          use Text::CSV_XS;

          my $migrator = shift;

          my $csv = Text::CSV_XS->new( ... );
          my $fh = IO::File->new( ... );

          my $sql = q[INSERT INTO Foo (size, description) VALUES (?, ?)];
          my $sth = $migrator->dbh()->prepare()

          while ( my $foo = $csv->getline_hr($fh) ) {
              $sth->execute( $foo->{size}, $foo->{description} );
          }

          $sth->finish();
      }

    The 02-add-bar-table/add-bar-table.sql file would contain DDL to create
    the "Bar" table.

HOW TO USE THIS DISTRIBUTION
    This distribution is not intended to be used all by itself. Instead, you
    will need to start with a DBMS-specific implementation like
    Database::Migrator::mysql.

    To actually run migrations, you either need to create an command line
    script or subclass an implementation (or both).

    The Database::Migrator::Core role consumes the MooseX::Getopt::Dashes
    role, making it easy to create a command line script for migrations:

      #!/usr/bin/env perl

      use strict;
      use warnings;

      use Database::Migrator::mysql;

      Database::Migrator::mysql->new_with_options()->create_or_update_database();

    However, all by itself, this will require quite a few command line
    options to be passed. You can simplify this by subclassing the
    implementation class and providing defaults for things like the
    migration directory and migration table.

    See the Database::Migrator::Core documentation for more details on what
    attributes you can provide defaults for.

THE APPLIED MIGRATION TABLE
    The fact that a migration has been applied is recorded in a table in the
    database.

    If you are creating a new schema from scratch, you can include this
    table. It should contain a single text column as its primary key. This
    column *must* be named "migration".

    The DDL to create this table might look like this:

      CREATE TABLE AppliedMigration (
          migration  TEXT  PRIMARY KEY
      );

  Bootstrapping This Table
    If you are migrating an existing schema to use this migration system,
    you will need to add this table to the schema. This can be done using
    the migration system itself. If the schema already exists but the table
    does not exist, it assumes that no migrations have been applied.

    In this case, you must ensure that the first migration adds this table.

      migrations/
       |
       |-- 00-add-applied-migration-table
       |    \
       |     \-- create-applied-migration-table.sql
       |
       |-- 01-add-foo-data/
       |     \
       |      \-- 01-create-foo-table.sql
       |      |
       |      |-- 02-insert-foo-data.pl
       |
       |-- 02-add-bar-table/
            \
             \-- add-bar-table.sql

    The 00-add-applied-migration-table/create-applied-migration-table.sql
    file would contain the DDL to create the table.

IDEMPOTENT MIGRATIONS
    Under normal operation, no migration should ever be applied twice.
    However, I still strongly recommend that you make all your migrations
    idempotent. This is much safer. For example, if the process applying
    migrations is killed, it's possible that it will be killed after a
    migration is applied but before that fact has been recorded.

SUPPORT
    Please submit bugs to the CPAN RT system at
    http://rt.cpan.org/NoAuth/Bugs.html?Dist=Database-Migrator or via email
    at bug-database-migrator@rt.cpan.org.

AUTHOR
    Dave Rolsky <autarch@urth.org>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2014 by MaxMind, Inc..

    This is free software, licensed under:

      The Artistic License 2.0 (GPL Compatible)



Changes
0.11     2014-04-08

- Changed the logging form Core so that when a database already exists this
  message is logged at the debug level, not info.


0.10     2014-02-06

- Previously, passing any one of username, password, host, or port from the
  command line would set the relevant attribute's value to "1", rather than
  what you passed. Reported by Lasse Makholm. RT #92726.

- The SQL used to insert a migration into the migration table now uses the
  column name. This allows you to put other columns in the table. Patch by
  Lasse Makholm. RT #92725.


0.09     2013-08-03

- Two migrations that started with the same number (01-foo and 01-bar) would
  not be sorted correctly. (But hopefully you never name things in such a
  confusing manner anyway). Patch by Reini Urban. RT #87205.


0.08     2013-07-18

- The Core role didn't load File::Slurp, which it needs to run Perl script
  migration files. Fixed by Olaf Alders.


0.07     2013-06-16

- Require implementations to provide the driver name rather than trying to
  guess it from the object's name. This was severely broken if you used a
  subclass that didn't end with something like "::mysql" or "::Pg".


0.06     2013-04-02

- Fixed some documentation errors.


0.05     2013-04-01

- Renamed the user attribute to username, but the former will still be
  supported by the constructor.

- Lots of refactorings to the Core role in preparation for
  Database::Migrator::Pg release.

- Added a class that helps with live tests - Test::Database::Migrator.


0.04     2012-04-22

- The logger attribute was not lazy, so it could be built before the quiet
  option was processed. This effectively meant that the quiet option was
  ignored.


0.03     2012-04-14

- Added missing prereqs on MooseX::Types and MooseX::Types::Path::Class.


0.02     2012-04-14

- Remove autodie from prereqs since there's nothing in the code that needs it
  right now.


0.01     2012-04-13

- First release upon an unsuspecting world.


lib/Test/Database/Migrator.pm
package Test::Database::Migrator;
$Test::Database::Migrator::VERSION = '0.11';
use strict;
use warnings;
use namespace::autoclean;

use File::Temp qw( tempdir );
use Database::Migrator::Types qw( ClassName Dir File Str );
use Log::Dispatch;
use Log::Dispatch::TestDiag;
use Path::Class qw( dir );
use Test::Fatal;
use Test::More 0.88;

use Moose;

has class => (
    is       => 'ro',
    isa      => ClassName,
    required => 1,
);

has database => (
    is      => 'ro',
    isa     => Str,
    default => 'DatabaseMigrator_test_' . $$,
);

has _tempdir => (
    is       => 'ro',
    isa      => Dir,
    init_arg => undef,
    lazy     => 1,
    default  => sub { dir( tempdir( CLEANUP => 1 ) ) },
);

has _schema_file => (
    is       => 'ro',
    isa      => File,
    init_arg => undef,
    lazy     => 1,
    default  => sub { $_[0]->_tempdir()->file('schema.sql') },
);

has _migrations_dir => (
    is       => 'ro',
    isa      => Dir,
    init_arg => undef,
    lazy     => 1,
    builder  => '_build_migrations_dir',
);

has _dbh => (
    is        => 'ro',
    isa       => 'DBI::db',
    init_arg  => undef,
    lazy      => 1,
    builder   => '_build_dbh',
    predicate => '_has_dbh',
);

sub run_tests {
    my $self = shift;

    $self->_write_ddl_file( $self->_schema_file(), $self->_schema_ddl() );

    ok(
        !$self->_new_migrator()->_database_exists(),
        $self->database() . ' does not exist yet'
    );

    $self->_check_initial_database();
    $self->_test_migrations();
}

sub _write_ddl_file {
    my $self = shift;
    my $file = shift;
    my $ddl  = shift;

    open my $fh, '>', $file;
    print {$fh} $ddl;
    close $fh;
}

sub _check_initial_database {
    my $self = shift;

    $self->_new_migrator()->create_or_update_database();

    my $migrator = $self->_new_migrator();

    ok(
        $migrator->_database_exists(),
        $self->database() . ' exists'
    );

    is_deeply(
        [ $self->_tables() ],
        [qw( applied_migration foo )],
        'newly created schema has the expected tables'
    );
}

sub _schema_ddl {
    return <<'EOF';
CREATE TABLE applied_migration (
    migration  VARCHAR(250)   PRIMARY KEY
);

CREATE TABLE foo (
    foo_id     INTEGER        PRIMARY KEY,
    foo_name   VARCHAR(50)    NOT NULL
);
EOF
}

sub _test_migrations {
    my $self = shift;

    $self->_write_first_migration();
    $self->_write_second_migration();

    $self->_new_migrator()->create_or_update_database();

    is_deeply(
        [ $self->_tables() ],
        [qw( applied_migration bar baz foo )],
        'migrated schema has the expected tables'
    );

    is_deeply(
        [ $self->_indexes_on('bar') ],
        ['bar_bar_name'],
        'bar table has the expected indexes'
    );

    is_deeply(
        [ $self->_indexes_on('baz') ],
        ['baz_baz_name'],
        'baz table has the expected indexes'
    );

    my $migrations = $self->_dbh()
        ->selectcol_arrayref('SELECT migration FROM applied_migration') || [];

    is_deeply(
        $migrations,
        [
            '01-first',
            '02-second',
        ],
        'migrations were recorded in the applied_migration table'
    );

    is(
        exception { $self->_new_migrator()->create_or_update_database() },
        undef,
        'no error running migrator again after migrations have been applied'
    );
}

# This migration writes out two separate files to test that the files in a
# given migration are run in sorted order. If they're not then the migration
# will fail entirely.
sub _write_first_migration {
    my $self = shift;

    my $dir = $self->_migrations_dir()->subdir('01-first');
    $dir->mkpath( 0, 0755 );

    my $create_tables = <<'EOF';
CREATE TABLE bar (
    bar_id     INTEGER        PRIMARY KEY,
    bar_name   VARCHAR(50)    NOT NULL
);

CREATE TABLE baz (
    baz_id     INTEGER        PRIMARY KEY,
    baz_name   VARCHAR(50)    NOT NULL
);
EOF

    $self->_write_ddl_file(
        $dir->file('01-create-tables.sql'),
        $create_tables,
    );

    $self->_write_ddl_file(
        $dir->file('02-create-bar-table-index.sql'),
        'CREATE INDEX bar_bar_name ON bar (bar_name)',
    );

    return;
}

sub _write_second_migration {
    my $self = shift;

    my $dir = $self->_migrations_dir()->subdir('02-second');
    $dir->mkpath( 0, 0755 );

    open my $fh, '>', $dir->file('01-create-baz-table-index.sql');
    print {$fh} <<'EOF';
CREATE INDEX baz_baz_name ON baz (baz_name);
EOF
    close $fh;

    return;
}

sub _new_migrator {
    my $self = shift;

    return $self->class()->new(
        database        => $self->database(),
        migration_table => 'applied_migration',
        schema_file     => $self->_schema_file(),
        migrations_dir  => $self->_migrations_dir(),
        logger          => _logger(),
    );
}

sub _logger {
    return Log::Dispatch->new(
        outputs => [ [ 'TestDiag', min_level => 'info' ] ],
    );
}

sub _build_migrations_dir {
    my $self = shift;

    my $dir = $self->_tempdir()->subdir('migrations');
    $dir->mkpath( 0, 0755 );

    return $dir;
}

sub _build_dbh {
    my $self = shift;

    return $self->_new_migrator()->dbh();
}

sub DEMOLISH {
    my $self = shift;

    if ( $ENV{DATABASE_MIGRATOR_TEST_WAIT} ) {
        print "\n  Waiting to clean up the test database\n\n";
        my $x = <STDIN>;
    }

    $self->_dbh()->disconnect() if $self->_has_dbh();

    $self->_new_migrator()->_drop_database();
}

=for Pod::Coverage .*

=cut

__PACKAGE__->meta()->make_immutable();

1;


lib/Database/Migrator.pm
package Database::Migrator;
$Database::Migrator::VERSION = '0.11';
# ABSTRACT: A system for implementing database migrations

__END__

=pod

=head1 NAME

Database::Migrator - A system for implementing database migrations

=head1 VERSION

version 0.11

=head1 DESCRIPTION

This distribution consists of a single role, L<Database::Migrator::Core>. This
role can be consumed by classes which implement the required methods for the
role. These classes will then implement a complete database schema creation
and migration system.

=head1 MIGRATION ARCHITECTURE

The migration system starts with a file containing the DDL (database
description language) for the full database schema. If the database doesn't
yet exist, the database will be created and this DDL will be run against it.

This DDL file I<should not> contain any sort of C<CREATE DATABASE>
statement. This will be done separately before the DDL is run.

This DDL file I<may> contain DDL to create users and grant them access to the
database.

Once the database exists, the migrations are run against the database.

Each migration goes into its own directory. The directory name is the name of
the migration. Migrations are applied in sorted order. If the migrations start
with numbers, they are sorted by these numbers, otherwise they are sorted
alphabetically.

The migration directory can either contain files with SQL or Perl. If a file
ends in ".sql", the migration runner code will feed it to the appropriate
command line utility for your database.

Otherwise the file is assumed to contain Perl code. This code is expected to
return a single anonymous subroutine when C<eval>ed. This subroutine will then
be called with the C<Database::Migrator> object as its only argument.

Each file in a single migration's directory is run in sorted order. You can
use numeric prefixes on these files if necessary.

Once a migration has been applied, that fact is stored in the database, and
the migration will not be applied again. This is done by recording the
migration's name in a table. The name of the table is determined by your
code. I recommend something like C<AppliedMigration> or C<applied_migrations>,
depending on your table naming scheme.

=head2 Migration Example

Let's assume a set of files like this:

  migrations/
   |
   |-- 01-add-foo-data/
   |     \
   |      \-- 01-create-foo-table.sql
   |      |
   |      |-- 02-insert-foo-data.pl
   |
   |-- 02-add-bar-table/
        \
         \-- add-bar-table.sql

The F<01-add-foo-data/01-create-foo-table.sql> file might look like this:

  CREATE TABLE Foo (
    foo_id       INT   PRIMARY KEY AUTO_INCREMENT,
    size         INT   NOT NULL,
    description  TEXT  NOT NULL
  );

The F<01-add-foo-data/02-insert-foo-data.pl> file might contain this:

  sub {
      use Text::CSV_XS;

      my $migrator = shift;

      my $csv = Text::CSV_XS->new( ... );
      my $fh = IO::File->new( ... );

      my $sql = q[INSERT INTO Foo (size, description) VALUES (?, ?)];
      my $sth = $migrator->dbh()->prepare()

      while ( my $foo = $csv->getline_hr($fh) ) {
          $sth->execute( $foo->{size}, $foo->{description} );
      }

      $sth->finish();
  }

The F<02-add-bar-table/add-bar-table.sql> file would contain DDL to create the
C<Bar> table.

=head1 HOW TO USE THIS DISTRIBUTION

This distribution is not intended to be used all by itself. Instead, you will
need to start with a DBMS-specific implementation like
L<Database::Migrator::mysql>.

To actually run migrations, you either need to create an command line script
or subclass an implementation (or both).

The L<Database::Migrator::Core> role consumes the L<MooseX::Getopt::Dashes>
role, making it easy to create a command line script for migrations:

  #!/usr/bin/env perl

  use strict;
  use warnings;

  use Database::Migrator::mysql;

  Database::Migrator::mysql->new_with_options()->create_or_update_database();

However, all by itself, this will require quite a few command line options to
be passed. You can simplify this by subclassing the implementation class and
providing defaults for things like the migration directory and migration
table.

See the L<Database::Migrator::Core> documentation for more details on what
attributes you can provide defaults for.

=head1 THE APPLIED MIGRATION TABLE

The fact that a migration has been applied is recorded in a table in the
database.

If you are creating a new schema from scratch, you can include this table. It
should contain a single text column as its primary key. This column I<must> be
named "migration".

The DDL to create this table might look like this:

  CREATE TABLE AppliedMigration (
      migration  TEXT  PRIMARY KEY
  );

=head2 Bootstrapping This Table

If you are migrating an existing schema to use this migration system, you will
need to add this table to the schema. This can be done using the migration
system itself. If the schema already exists but the table does not exist, it
assumes that no migrations have been applied.

In this case, you must ensure that the first migration adds this table.

  migrations/
   |
   |-- 00-add-applied-migration-table
   |    \
   |     \-- create-applied-migration-table.sql
   |
   |-- 01-add-foo-data/
   |     \
   |      \-- 01-create-foo-table.sql
   |      |
   |      |-- 02-insert-foo-data.pl
   |
   |-- 02-add-bar-table/
        \
         \-- add-bar-table.sql

The F<00-add-applied-migration-table/create-applied-migration-table.sql> file
would contain the DDL to create the table.

=head1 IDEMPOTENT MIGRATIONS

Under normal operation, no migration should ever be applied twice. However, I
still strongly recommend that you make all your migrations idempotent. This is
much safer. For example, if the process applying migrations is killed, it's
possible that it will be killed after a migration is applied but before that
fact has been recorded.

=head1 SUPPORT

Please submit bugs to the CPAN RT system at
http://rt.cpan.org/NoAuth/Bugs.html?Dist=Database-Migrator or via email at
bug-database-migrator@rt.cpan.org.

=head1 AUTHOR

Dave Rolsky <autarch@urth.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2014 by MaxMind, Inc..

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut


lib/Database/Migrator/Core.pm
package Database::Migrator::Core;
$Database::Migrator::Core::VERSION = '0.11';
use strict;
use warnings;
use namespace::autoclean;

use Database::Migrator::Types qw( ArrayRef Bool Dir File Maybe Str );
use DBI;
use Eval::Closure qw( eval_closure );
use File::Slurp qw( read_file );
use Log::Dispatch;
use Moose::Util::TypeConstraints qw( duck_type );

use Moose::Role;

with 'MooseX::Getopt::Dashes';

MooseX::Getopt::OptionTypeMap->add_option_type_to_map(
    Maybe [Str] => '=s',
);

requires qw(
    _create_database
    _driver_name
    _drop_database
    _run_ddl
);

has database => (
    is       => 'ro',
    isa      => Str,
    required => 1,
);

has [qw( username password host port )] => (
    is      => 'ro',
    isa     => Maybe [Str],
    default => undef,
);

has migration_table => (
    is       => 'ro',
    isa      => Str,
    required => 1,
);

has migrations_dir => (
    is       => 'ro',
    isa      => Dir,
    coerce   => 1,
    required => 1,
);

has schema_file => (
    is       => 'ro',
    isa      => File,
    coerce   => 1,
    required => 1,
);

has _database_exists => (
    is       => 'ro',
    isa      => Bool,
    init_arg => undef,
    lazy     => 1,
    builder  => '_build_database_exists',
);

has __pending_migrations => (
    traits   => ['Array'],
    is       => 'ro',
    isa      => ArrayRef [Dir],
    init_arg => undef,
    lazy     => 1,
    builder  => '_build_pending_migrations',
    handles  => {
        _pending_migrations    => 'elements',
        has_pending_migrations => 'count',
    },
);

has dbh => (
    traits   => ['NoGetopt'],
    is       => 'ro',
    isa      => 'DBI::db',
    init_arg => undef,
    lazy     => 1,
    builder  => '_build_dbh',
);

has logger => (
    traits  => ['NoGetopt'],
    is      => 'ro',
    isa     => duck_type( [qw( debug info )] ),
    lazy    => 1,
    builder => '_build_logger',
);

has verbose => (
    is      => 'ro',
    isa     => Bool,
    default => 0,
);

has quiet => (
    is      => 'ro',
    isa     => Bool,
    default => 0,
);

has dry_run => (
    is      => 'ro',
    isa     => Bool,
    default => 0,
);

around BUILDARGS => sub {
    my $orig  = shift;
    my $class = shift;

    my $p = $class->$orig(@_);

    $p->{username} = delete $p->{user}
        if exists $p->{user};

    return $p;
};

sub BUILD { }
after BUILD => sub {
    my $self = shift;

    die 'Cannot be both quiet and verbose'
        if $self->quiet() && $self->verbose();
};

sub create_or_update_database {
    my $self = shift;

    if ( $self->_database_exists() ) {
        my $database = $self->database();
        $self->logger()->debug("The $database database already exists");
    }
    else {
        $self->_create_database();
        $self->_run_ddl( $self->schema_file() );
    }

    $self->_run_migrations();

    return;
}

sub _run_migrations {
    my $self = shift;

    $self->_run_one_migration($_) for $self->_pending_migrations();
}

sub _run_one_migration {
    my $self      = shift;
    my $migration = shift;

    my $name = $migration->basename();

    $self->logger()->info("Running migration - $name");

    my @files = grep { !$_->is_dir() } $migration->children( no_hidden => 1 );

    for my $file ( sort _numeric_or_alpha_sort @files ) {
        my $basename = $file->basename();
        if ( $file =~ /\.sql/ ) {
            $self->logger()->debug(" - running $basename as sql");
            $self->_run_ddl($file);
        }
        else {
            $self->logger()->debug(" - running $basename as perl code");

            my $perl = read_file( $file->stringify() );

            my $sub = eval_closure( source => $perl );

            next if $self->dry_run();

            $sub->($self);
        }
    }

    return if $self->dry_run();

    my $table = $self->dbh()->quote_identifier( $self->migration_table() );
    $self->dbh()->do( "INSERT INTO $table (migration) VALUES (?)", undef, $name );

    return;
}

sub _build_pending_migrations {
    my $self = shift;

    my $table = $self->migration_table();

    my %ran;
    if ( grep { $_ =~ /\b\Q$table\E\b/ } $self->dbh()->tables() ) {
        my $quoted = $self->dbh()->quote_identifier($table);

        %ran
            = map { $_ => 1 }
            @{ $self->dbh()
                ->selectcol_arrayref("SELECT migration FROM $quoted") || [] };
    }

    return [
        sort _numeric_or_alpha_sort grep { !$ran{ $_->basename() } }
            grep                         { $_->is_dir() }
            $self->migrations_dir()->children( no_hidden => 1 )
    ];
}

sub _build_logger {
    my $self = shift;

    my $outputs
        = $self->quiet()
        ? [ 'Null', min_level => 'emerg' ]
        : [
        'Screen',
        min_level => ( $self->verbose() ? 'debug' : 'info' ),
        newline => 1,
        ];

    return Log::Dispatch->new( outputs => [$outputs] );
}

sub _build_database_exists {
    my $self = shift;

    local $@;
    return eval { $self->_build_dbh() } ? 1 : 0;
}

sub _build_dbh {
    my $self = shift;

    return DBI->connect(
        'dbi:' . $self->_driver_name() . ':database=' . $self->database(),
        $self->username(),
        $self->password(),
        {
            RaiseError         => 1,
            PrintError         => 1,
            PrintWarn          => 1,
            ShowErrorStatement => 1,
        },
    );
}

sub _numeric_or_alpha_sort {
    my ( $a_num, $a_alpha ) = $a->basename() =~ /^(\d+)(.+)/;
    my ( $b_num, $b_alpha ) = $b->basename() =~ /^(\d+)(.+)/;

    $a_num ||= 0;
    $b_num ||= 0;

    return ( $a_num <=> $b_num or $a_alpha cmp $b_alpha );
}


1;

# ABSTRACT: Core role for Database::Migrator implementation classes

__END__

=pod

=head1 NAME

Database::Migrator::Core - Core role for Database::Migrator implementation classes

=head1 VERSION

version 0.11

=head1 SYNOPSIS

  package Database::Migrator::SomeDB;

  use Moose;
  with 'Database::Migrator::Core';

  sub _build_database_exists { ... }
  sub _build_dbh             { ... }
  sub _create_database       { ... }

=head1 DESCRIPTION

This role implements the bulk of the migration logic, leaving a few details up
to DBMS-specific classes.

You can then subclass these DBMS-specific classes to provide defaults for
various attributes, or to override some of the implementation.

=for Pod::Coverage   BUILD
  create_or_update_database

=head1 PUBLIC ATTRIBUTES

This role defines the following public attributes. These attributes may be
provided via the command line or you can set defaults for them in a subclass.

=over 4

=item * database

The name of the database that will be created or migrated. This is required.

=item * username, password, host, port

These parameters are used when connecting to the database. They are all
optional.

=item * migration_table

The name of the table which stores the name of applied migrations. This is
required.

=item * migrations_dir

The directory containing migrations. This is required, but it is okay if the
directory is empty.

=item * schema_file

The full path to the file containing the initial schema for the database. This
will be used to create the database if it doesn't already exist. This is required.

=item * verbose

This affects the verbosity of output logging. Defaults to false.

=item * quiet

If this is true, then no output will logged at all. Defaults to false.

=item * dry_run

If this is true, no migrations are actually run. Instead, the code just logs
what it I<would> do. Defaults to false.

=back

=head1 METHODS

This role provide just one public method, C<create_or_update_database()>.

It will create a new database if none exists.

It will run all unapplied migrations on this schema once it does exist.

=head1 REQUIRED METHODS

If you want to create your own implementation class, you must implement the
following methods. All of these methods should throw an error

=head2 $migration->_create_database()

This should create an I<empty> database. This role will take care of executing
the DDL for defining the schema.

=head2 $migration->_driver_name()

This return a string containing the DBI driver name, such as "mysql" or "Pg".

=head2 $migration->_drop_database()

This should drop the database. Right now it is only used for testing.

=head2 $migration->_run_ddl($ddl)

Given a string containing one or more DDL statements, this method must run
that DDL against the database.

=head1 OVERRIDEABLE ATTRIBUTES AND METHODS

There are a number of attributes methods in this role that you may wish to
override in a custom subclass of an implementation.

For any attribute where you provide a default value, make sure to also set C<<
required => 0 >> as well.

=over 4

=item * database attribute

You can provide a default database name.

=item * username, password, host, and port attributes

You can provide a default values for these connection attributes.

=item * migration_table

You can provide a default table name.

=item * migrations_dir

You can provide a default directory.

=item * schema_file

You can provide a default file name.

=item * _build_logger()

You must return an object with C<debug()> and C<info()> methods.

=back

=head1 AUTHOR

Dave Rolsky <autarch@urth.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2014 by MaxMind, Inc..

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.015.
use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A system for implementing database migrations",
  "AUTHOR" => "Dave Rolsky <autarch\@urth.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Database-Migrator",
  "EXE_FILES" => [],
  "LICENSE" => "artistic_2",
  "NAME" => "Database::Migrator",
  "PREREQ_PM" => {
    "DBI" => 0,
    "Eval::Closure" => 0,
    "File::Slurp" => 0,
    "File::Temp" => 0,
    "Log::Dispatch" => 0,
    "Log::Dispatch::TestDiag" => 0,
    "Moose" => 0,
    "Moose::Role" => 0,
    "Moose::Util::TypeConstraints" => 0,
    "MooseX::Getopt::Dashes" => 0,
    "MooseX::Types" => 0,
    "MooseX::Types::Combine" => 0,
    "MooseX::Types::Moose" => 0,
    "MooseX::Types::Path::Class" => 0,
    "Path::Class" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.88",
    "namespace::autoclean" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Find" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "0.11",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "DBI" => 0,
  "Eval::Closure" => 0,
  "File::Find" => 0,
  "File::Slurp" => 0,
  "File::Temp" => 0,
  "Log::Dispatch" => 0,
  "Log::Dispatch::TestDiag" => 0,
  "Moose" => 0,
  "Moose::Role" => 0,
  "Moose::Util::TypeConstraints" => 0,
  "MooseX::Getopt::Dashes" => 0,
  "MooseX::Types" => 0,
  "MooseX::Types::Combine" => 0,
  "MooseX::Types::Moose" => 0,
  "MooseX::Types::Path::Class" => 0,
  "Path::Class" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "0.88",
  "namespace::autoclean" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



